### ISSUE TITLE
"VMC" vs "VMCBatched" GPU version energies differ.

### ISSUE THREAD

**Issue by aannabe** [QMCPACK/qmcpack#3496]
**Describe the bug**
Total energies differ when using the GPU version of the code with `vmc` vs `vmc_batch` drivers.
The inputs are identical except for some changes in the walker counts.
Using `vmc` driver:
```
  LocalEnergy           =        -2233.5973 +/-           0.0020    1.0 
  Variance              =            19.174 +/-            0.032    1.3 
  Kinetic               =           837.337 +/-            0.076    2.5 
  LocalPotential        =         -3070.934 +/-            0.076    2.5 
  AcceptRatio           =          0.505145 +/-         0.000020    2.9 
```
Using `vmc_batch` driver:
```
  LocalEnergy           =        -2233.6471 +/-           0.0073    2.1 
  Variance              =            19.600 +/-            0.041    1.3 
  Kinetic               =           837.204 +/-            0.045    1.0 
  LocalPotential        =         -3070.851 +/-            0.046    1.0 
  AcceptRatio           =          0.505153 +/-         0.000012    1.4 
```
I have previously run tests on the same system for CPU version vs non-batched GPU version and they agreed.
So I think that the non-batched energy is the correct value.

**To Reproduce**
Both runs use the below version of the code:
```
  Git branch: update-summit-recipes
  Last git commit: 4cdbfa380f4e17e5923631c765696631ac2d7c70-dirty
  Last git commit date: Mon Aug 23 19:27:10 2021 -0400
  Last git commit subject: Update Summit build recipes.
```
Inputs and outputs can be found at:
[issue_batch.tar.gz](https://github.com/QMCPACK/qmcpack/files/7261311/issue_batch.tar.gz)

**Expected behavior**
VMC energies should agree.

**Machine:**
 - OLCF/Summit
 - Loaded modules and job details are in the attached tar file.

**System/Material**
- Charged (anion) solid RuCl3 in 2x1x2 conventional cell (593 electrons). Orbitals are from Espresso.

**Comment by prckent** [comment]
Please can you update the issue to say what the system that you are simulating is?

**Comment by aannabe** [comment]
Edited to give more info

**Comment by ye-luo** [comment]
I'm testing a fix to non-deterministic behaviors in offload J2. Will test your case.

**Comment by prckent** [comment]
Once this is understood we will need to add a small reproducer test to avoid anything similar.

**Comment by prckent** [comment]
It would be interesting to know if the problem does not occur with no Jastrow present.

**Comment by aannabe** [comment]
I was setting up the no-Jastrow runs but it seems it is not needed now. Using Ye's new executable below:
```
  Git branch: opt-J2
  Last git commit: 4431991df2450bce3c9e1dd62f0c57319d091f35-dirty
  Last git commit date: Thu Sep 30 23:23:17 2021 -0400
  Last git commit subject: Merge branch cudatoolkit into opt-J2
```
the `vmc` and `vmc_batch` energies agree within the error bars for the above-mentioned system.
`vmc` driver energy:
```
gs_dmc  series 0 
  LocalEnergy           =        -2233.5940 +/-           0.0021    1.0 
  Variance              =            19.159 +/-            0.035    1.3 
  Kinetic               =           837.229 +/-            0.022    1.0 
  LocalPotential        =         -3070.823 +/-            0.022    1.0 
  AcceptRatio           =         0.5051672 +/-        0.0000039    1.0 
```
`vmc_batch` driver energy:
```
gs_dmc  series 0 
  LocalEnergy           =        -2233.5963 +/-           0.0010    1.0 
  Variance              =            19.159 +/-            0.020    1.0 
  Kinetic               =           837.200 +/-            0.018    1.0 
  LocalPotential        =         -3070.796 +/-            0.018    1.0 
  AcceptRatio           =         0.5051734 +/-        0.0000035    1.0 
```
These inputs and outputs are attached here:
[resolved_issue.tar.gz](https://github.com/QMCPACK/qmcpack/files/7269269/resolved_issue.tar.gz)

**Comment by prckent** [comment]
Fixed by #3498
